1178 robin conditions in models

IngridKJ opened the pull request:
## Proposed changes

Addresses issue #1178: Allow for usage of Robin type boundary conditions with the model class setups in PorePy.

This PR includes adding a robin operator parameter to the method _combine_boundary_operators(). In addition to this, each call to _combine_boundary_operators() is refactored into new methods specific for each physical problem, distinguished by their flux name:
* combine_boundary_operators_darcy_flux()
* combine_boundary_operators_fourier_flux()
* combine_boundary_operators_mechanical_stress()

Additionally: Fixes a problem of singular matrices related to nodes with only one subcell gradient but two or more boundary faces (for Robin conditions. This was already in place for Neumann). 

### Practicalities
The Robin conditions are implemented in the following form in PorePy: flux + alpha * potential = G. 
What is needed by the user to utilize such boundary conditions is to: 
In bc_type-methods: 
* Use the flag "rob" when setting boundary condition type, analogous to setting Dirichlet ("dir") and Neuman ("neu").
* Assign the Robin weight, alpha. The Robin weight is an attribute of the boundary condition class (.robin_weight). See the `__init__`-methods of BoundaryCondition/BoundaryConditionVectorial (porepy/params/bc.py) for an example of what the weight should look like.

In bc_values-methods:
* Assign the right hand side G. G has units of flux, and is set by bc_values_"flux_name"() - same as the Neumann values. As both BC type values are set by the same method, it is important to make sure the Neumann and Robin values are assigned to the correct boundary sides.

Why refactoring the combine operators call? 
Tailoring of each operator is now easily achieved by just overriding the wrapper method (combine_boundary_operators_"flux_name"). Note: The default Robin operator is the same as the Neumann operator.

## Types of changes

What types of changes does this PR introduce to PorePy?
_Put an `x` in the boxes that apply._

- [ ] Minor change (e.g., dependency bumps, broken links).
- [ ] Bugfix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [x] Testing (contribution related to testing of existing or new functionality).
- [ ] Documentation (contribution related to adding, improving, or fixing documentation).
- [ ] Maintenance (e.g., improve logic and performance, remove obsolete code).
- [ ] Other:

## Checklist

_Put an `x` in the boxes that apply or explain briefly why the box is not relevant._

- [x] The documentation is up-to-date.
- [x] Static typing is included in the update.
- [x] This PR does not duplicate existing functionality.
- [ ] The update is covered by the test suite (including tests added in the PR).
- [ ] If new skipped tests have been introduced in this PR, `pytest` was run with the `--run-skipped` flag.

OmarDuran reviewed (changes requested):
Thank you for your contribution. 

It is important to consider the following:

- Perform a test to ensure that boundary segments labeled with Robin type do not intersect with other types;

- Include a test for scalar and vector problem, for which Robin approximates the functionality of Dirichlet and Neumann;

- Include a tutorial that illustrates how the boundary condition is intended to be used in the models.

OmarDuran commented on src/porepy/applications/test_utils/models.py:
```diff
@@ -118,6 +118,87 @@ class Model(geometry, model_class):
     return model
 
 
+class ThreeBoundaryConditionTypes:
```
Consider renaming it to `BoundaryConditionTypes`.

OmarDuran commented on tests/models/test_thermoporomechanics.py:
```diff
@@ -265,6 +265,77 @@ def test_positive_p_frac_positive_opening():
     assert np.allclose(t_frac, 8.3e-6, atol=1e-7)
 
 
+def test_robin_boundary_flux():
```
For clarity, consider explicit assignment of the `robin_weight` parameter.

OmarDuran commented:
The question is general in nature. I believe that there is room for a unification of names and types, such as

Types:
- Dirichlet_type
- Neumann_type
- Robin_type

Data:
- Dirichlet_data
- Neumann_data
- Robin_data

IvarStefansson commented:
> Thank you for your contribution.
> 
> It is important to consider the following:
> 
> * Perform a test to ensure that boundary segments labeled with Robin type do not intersect with other types;
> * Included is a test for a scalar and vector problem, for which Robin approximates the functionality of Dirichlet and Neumann;
> * Include a tutorial that illustrates how the boundary condition are intended to be used in the models.

To your first point, @OmarDuran, isn't that more a test for the `BoundaryCondition` class, rather than this Model integration?

IvarStefansson commented on tests/models/test_boundary_condition.py:
```diff
@@ -105,3 +105,204 @@ def test_boundary_condition_mixin(t_end: int):
+
+    def bc_values_pressure(self, bg: pp.BoundaryGrid) -> np.ndarray:
+        """Assigns pressure values on the west boundary."""
+        values = np.zeros(bg.num_cells)
```
Wouldn't it be even better to test for nonzero values?

IngridKJ replied:
So, "more nonzero" than what is currently in place? It is technically nonzero now in the sense that I have unit values on the west boundary.

IvarStefansson replied:
Yes, but that's only the Dirichlet one, not the ones you're actually testing, right?

IngridKJ replied:
The values I am testing are values of primary the variables in all the domain cells. So even though the boundary values are zero, the values I am testing are not. But if nonzero also on the other boundaries make more sense, I can change it.

IvarStefansson commented on tests/models/test_boundary_condition.py:
```diff
@@ -105,3 +105,204 @@ def test_boundary_condition_mixin(t_end: int):
+"""Here follows mixins related to testing of Robin limit cases, and eventually the test itself. """
+
+
+class Geometry:
```
I think we can use `SquareDomainOrthogonalFractures` with the parameter `fracture_indices` equal to `[]`.

IngridKJ commented:
@IvarStefansson: PR is ready for (hopefully final) review now. I addressed your comment about nonzero boundary values, and after the meeting with you and Eirik today, I have also refactored the eliminate-asymmetric-stress-tensor functionality for Neumann/Robin.

IvarStefansson commented on src/porepy/models/constitutive_laws.py:
```diff
@@ -1765,6 +1776,34 @@ def potential_trace(
         )
         return pressure_trace
 
+    def combine_boundary_operators_tpfa_flux(
```
I think we need access to the separate methods for Darcy and Fourier, probably passed as arguments to diffusive_flux. Let's discuss if you disagree.

IvarStefansson commented on src/porepy/models/constitutive_laws.py:
```diff
@@ -2715,6 +2745,29 @@ def fourier_flux(self, subdomains: list[pp.Grid]) -> pp.ad.Operator:
+    def combine_boundary_operators_fourier_flux(
+        self, subdomains: list[pp.Grid]
+    ) -> pp.ad.Operator:
+        """Combine fourier flux boundary operators.
```
Please capitalize Fourier and Darcy throughout.

IvarStefansson commented on src/porepy/models/constitutive_laws.py:
```diff
@@ -3428,6 +3474,30 @@ def mechanical_stress(self, domains: pp.SubdomainsOrBoundaries) -> pp.ad.Operato
+    def combine_boundary_operators_mechanical_stress(
+        self, subdomains: list[pp.Grid]
+    ) -> pp.ad.Operator:
+        """Combine mechanical stress boundary operators.
```
Should we explain in the documentation of these methods that they can be used to define other Robin operators? I.e., not use `mechanical_stress` for both neu and rob.

IvarStefansson commented on src/porepy/models/energy_balance.py:
```diff
@@ -355,6 +355,7 @@ def enthalpy_dirichlet(boundary_grids):
                 subdomains=subdomains,
                 dirichlet_operator=enthalpy_dirichlet,
                 neumann_operator=self.enthalpy_flux,
+                robin_operator=self.enthalpy_flux,
```
Are you sure this is relevant for advective fluxes? I don't think Upwind handles Robin.

IngridKJ replied:
Not sure, no. I do not know nearly enough about this, so when I went through the code after the Robin-relevant changes, I just adjusted all ``_combine_boundary_ooperators`` calls to match the new signature. If it is not relevant for certain fluxes, maybe the robin_operator in ``_combine_boundary_ooperators`` can be optional to pass?

IvarStefansson commented on src/porepy/numerics/fv/mpsa.py:
```diff
@@ -832,8 +832,23 @@ def _create_inverse_gradient_matrix(
         # an inconsistency.
-        self._eliminate_ncasym_neumann(
-            ncasym, subcell_topology, bound_exclusion, cell_node_blocks, nd
+        self._eliminate_ncasym(
```
Great catch!

IvarStefansson commented on tests/models/test_boundary_condition.py:
```diff
@@ -102,3 +108,205 @@ def test_boundary_condition_mixin(t_end: int):
+        bc.robin_weight = np.reshape(r_w, (sd.dim, sd.dim, sd.num_faces), "F") * 0
+        return bc
+
+    def bc_type_scalar_rob_dir(self, sd: pp.Grid) -> pp.BoundaryCondition:
```
If you rename this method to e.g. `bc_type_scalar`, you can let the below class inherit from this, avoiding redefinition of `bc_type_x_flux` methods.

IngridKJ replied:
Maybe I misunderstand something here, but I do not see how that will help the case. The bc_type_scalar methods are different (one assigns Neumann, one assigns Robin), so redefinition is needed in some sense anyways? Or?

IvarStefansson commented on tests/models/test_boundary_condition.py:
```diff
@@ -102,3 +108,205 @@ def test_boundary_condition_mixin(t_end: int):
+    "balance_type", ["mass_balance", "momentum_balance", "mass_and_energy_balance"]
+)
+def test_robin_limit_case(run_models, balance_type):
+    """Test that Robin conditions are equivalent to Neumann with Robin weight = 0.
```
Nice documentation. Thanks!

OmarDuran replied:
In case I am not mistaken, you are testing one limit only, alpha -> 0 (Neuman). However, I believe that it would be relevant to include a test for alpha -> inf (Dirichlet), numerically large values {1e+12, 1e+14, 1e+16}.

IngridKJ replied:
Makes sense to include both. I am not sure if I will be able to finish it before I go off on holidays, but I have it on my to do list.

IvarStefansson commented on src/porepy/applications/test_utils/models.py:
```diff
@@ -118,6 +118,91 @@ class Model(geometry, model_class):
     return model
 
 
+class BoundaryConditionTypes:
```
Please comment on whether this class has overlap with existing ones.

IngridKJ replied:
Not that I could find, no.

IvarStefansson reviewed (commented):
Good work! Some relatively minor comments.
